indexer-core/src/main/java/org/apache/maven/index/updater/IndexDataReader.java

```diff
@@ -277,17 +277,17 @@ private IndexWriter tempWriter( final String name ) throws IOException
     }
 
     private void addToIndex( final Document doc, final IndexingContext context, final IndexWriter indexWriter,
-                             final ConcurrentMap<String, Boolean> rootGroups,
-                             final ConcurrentMap<String, Boolean> allGroups )
+                             final Set<String> rootGroups,
+                             final Set<String> allGroups )
             throws IOException
     {
         ArtifactInfo ai = IndexUtils.constructArtifactInfo( doc, context );
         if ( ai != null )
         {
             indexWriter.addDocument( IndexUtils.updateDocument( doc, context, false, ai ) );
 
-            rootGroups.putIfAbsent( ai.getRootGroup(), Boolean.TRUE );
-            allGroups.putIfAbsent( ai.getGroupId(), Boolean.TRUE );
+            rootGroups.add( ai.getRootGroup() );
+            allGroups.add( ai.getGroupId() );
```

Comment:
`add` should be a little bit faster compared to `putIfAbsent` in the concurrent case. But since so much IO is going on it goes down in the noise.